File: app/src/main/java/com/example/oblivion/GameSurfaceView.java
package com.example.oblivion;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;

public class GameSurfaceView extends GLSurfaceView {

    private static final String TAG = "GameSurfaceView";
    private GameRenderer renderer;
    private float lastX = 0;
    private float lastY = 0;
    private boolean renderThreadInitialized = false;

    public GameSurfaceView(Context context) {
        super(context);
        Log.i(TAG, "Constructor(context) called");
        init();
    }

    public GameSurfaceView(Context context, AttributeSet attrs) {
        super(context, attrs);
        Log.i(TAG, "Constructor(context, attrs) called");
        init();
    }

    private void init() {
        Log.i(TAG, "=== init() called ===");
        try {
            // OpenGL ES 3.0 を使用
            Log.i(TAG, "Setting EGL context client version to 3");
            setEGLContextClientVersion(3);

            // GameRenderer を設定
            Log.i(TAG, "Creating GameRenderer");
            renderer = new GameRenderer(this);
            Log.i(TAG, "Setting renderer");
            setRenderer(renderer);
            Log.i(TAG, "Renderer set successfully");

            // フレームレート制御: 連続レンダリングを使用し、native側で frame timing を実装
            // Continuous rendering allows native code to control frame rate precisely
            Log.i(TAG, "Setting render mode to RENDERMODE_CONTINUOUSLY");
            setRenderMode(GLSurfaceView.RENDERMODE_CONTINUOUSLY);

            // OPTIMIZATION: Disabled setPreserveEGLContextOnPause for Amazon Fire compatibility
            // This was causing render thread initialization failures on older devices
            // Log.i(TAG, "Setting preserve EGL context on pause");
            // setPreserveEGLContextOnPause(true);

            Log.i(TAG, "=== init() completed successfully ===");
        } catch (Exception e) {
            Log.e(TAG, "Exception in init: " + e.getMessage(), e);
            Log.e(TAG, "Stack trace: ", e);
            // Continue anyway - don't fail initialization
        }
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        float x = event.getX();
        float y = event.getY();

        switch (event.getAction()) {
            case MotionEvent.ACTION_MOVE:
                float dx = x - lastX;
                float dy = y - lastY;

                // カメラ制御にタッチイベントを伝達
                if (renderer != null) {
                    renderer.onTouchEvent(dx, dy);
                }
                break;
        }

        lastX = x;
        lastY = y;
        return true;
    }

    @Override
    public void onResume() {
        Log.i(TAG, "onResume called");
        try {
            Log.i(TAG, "Resuming renderer");
            renderThreadInitialized = false;
            super.onResume();
            Log.i(TAG, "onResume completed, render thread queued");
        } catch (Exception e) {
            Log.e(TAG, "Exception in onResume: " + e.getMessage(), e);
        }
    }

    @Override
    public void onPause() {
        Log.i(TAG, "onPause called");
        try {
            Log.i(TAG, "Pausing renderer");
            super.onPause();
            Log.i(TAG, "onPause completed");
        } catch (Exception e) {
            Log.e(TAG, "Exception in onPause: " + e.getMessage(), e);
        }
    }

    public void setRenderThreadInitialized(boolean initialized) {
        renderThreadInitialized = initialized;
        Log.i(TAG, "Render thread initialized flag: " + initialized);
    }

    public GameRenderer getGameRenderer() {
        return renderer;
    }
}

File: app/src/main/java/com/example/oblivion/GameSurfaceView.kt
package com.example.oblivion

import android.content.Context
import android.opengl.GLSurfaceView
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent

class GameSurfaceView : GLSurfaceView {

    private var renderer: GameRenderer? = null
    private var renderThreadInitialized = false
    private var rendererSet = false

    // GL viewport dimensions (set by renderer after surface creation)
    private var glViewportWidth = 0
    private var glViewportHeight = 0

    // Display density for touch coordinate scaling
    // MotionEvent.getX()/getY() returns coordinates in display logical pixels,
    // but the UI (GL viewport) uses physical pixels. We need to scale.
    private var displayDensity = 1.0f

    companion object {
        private const val TAG = "GameSurfaceView"
    }

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    private fun init() {
        try {
            setEGLContextClientVersion(3)
            val newRenderer = GameRenderer(this)
            renderer = newRenderer
            // Defer setRenderer to onResume to ensure GL thread starts after activity is visible
            Log.i(TAG, "GameSurfaceView initialized (renderer deferred)")

            // Get display density for touch coordinate scaling
            displayDensity = context.resources.displayMetrics.density
            Log.i(TAG, "Display density: $displayDensity")
        } catch (e: Exception) {
            Log.e(TAG, "Exception in init: ${e.message}", e)
        }
    }

    /**
     * Call this after setContentView and onResume to start the GL render thread.
     */
    fun startRenderer() {
        if (!rendererSet && renderer != null) {
            Log.i(TAG, "Starting GL renderer thread")
            setRenderer(renderer)
            renderMode = RENDERMODE_CONTINUOUSLY
            rendererSet = true
        }
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        Log.d(TAG, "dispatchTouchEvent: action=${event.actionMasked}")
        return super.dispatchTouchEvent(event)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        if (renderer == null) {
            Log.w(TAG, "onTouchEvent: renderer is null!")
            return false
        }

        // MotionEvent.getX()/getY() returns coordinates in the view's physical pixel space.
        // For a match_parent GLSurfaceView, this already matches the GL viewport (physical pixels).
        // No density scaling needed - coordinates are already in the correct space.
        val action = event.actionMasked
        val actionIndex = event.actionIndex
        
        when (action) {
            MotionEvent.ACTION_DOWN, MotionEvent.ACTION_POINTER_DOWN -> {
                val pointerId = event.getPointerId(actionIndex)
                val x = event.getX(actionIndex)
                val y = event.getY(actionIndex)
                Log.d(TAG, "Touch DOWN: pointerId=$pointerId, pos=($x, $y)")
                renderer?.onTouchEvent(pointerId, x, y, 0) // 0 = DOWN
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_POINTER_UP -> {
                val pointerId = event.getPointerId(actionIndex)
                val x = event.getX(actionIndex)
                val y = event.getY(actionIndex)
                Log.d(TAG, "Touch UP: pointerId=$pointerId, pos=($x, $y)")
                renderer?.onTouchEvent(pointerId, x, y, 1) // 1 = UP
            }
            MotionEvent.ACTION_MOVE -> {
                for (i in 0 until event.pointerCount) {
                    val pointerId = event.getPointerId(i)
                    val x = event.getX(i)
                    val y = event.getY(i)
                    renderer?.onTouchEvent(pointerId, x, y, 2) // 2 = MOVE
                }
            }
            MotionEvent.ACTION_CANCEL -> {
                // Release all active pointers on cancel
                for (i in 0 until event.pointerCount) {
                    val pointerId = event.getPointerId(i)
                    val x = event.getX(i)
                    val y = event.getY(i)
                    renderer?.onTouchEvent(pointerId, x, y, 1) // 1 = UP
                }
                Log.d(TAG, "Touch CANCEL: released all pointers")
            }
        }
        return true
    }

    override fun onResume() {
        // If renderer not yet set, start it now (before super.onResume)
        if (!rendererSet && renderer != null) {
            startRenderer()
        }
        renderThreadInitialized = false
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
    }

    fun setRenderThreadInitialized(initialized: Boolean) {
        renderThreadInitialized = initialized
    }

    /**
     * Called by the renderer after surface creation to inform the view
     * of the actual GL viewport dimensions. Used for touch coordinate scaling.
     */
    fun setGLViewportSize(width: Int, height: Int) {
        glViewportWidth = width
        glViewportHeight = height
        Log.i(TAG, "GL viewport size set: ${width}x${height}, view size: ${this.width}x${this.height}")
    }

    fun getGameRenderer(): GameRenderer? {
        return renderer
    }
}
